R/methods-MbioDataset.R

```diff
@@ -0,0 +1,206 @@
+#' @importFrom phyloseq phyloseq
+#' @importFrom microbiomeComputations AbundanceData
+#' @rdname getCollection
+#' @export
+setGeneric("getCollection", function(object, collectionName, format = c("AbundanceData", "phyloseq", "Collection"), continuousMetadataOnly = c(FALSE, TRUE)) standardGeneric("getCollection"))
+
+#' @rdname getCollection
+#' @aliases getCollection,MbioDataset,character-method
+setMethod("getCollection", "MbioDataset", function(object, collectionName = character(0), format = c("AbundanceData", "phyloseq", "Collection"), continuousMetadataOnly = c(FALSE, TRUE)) {
+    format <- veupathUtils::matchArg(format)
+    continuousMetadataOnly <- veupathUtils::matchArg(continuousMetadataOnly)
+    
+    if (length(collectionName) == 0) {
+        stop("Must specify a collection name")
+    }
+    
+    if (!collectionName %in% getCollectionNames(object)) {
+        stop(sprintf("Collection '%s' does not exist", collectionName))
+    }
+
+    collection <- object@collections[[which(getCollectionNames(object) == collectionName)]]
+    if (format == "Collection") {
+        return(collection)
+    }
+
+    collectionDT <- data.table::setDT(collection@data)
+    collectionIdColumns <- c(collection@recordIdColumn, collection@ancestorIdColumns)
+
+    # remove IRI from collection column names. strip everything after and including the last square bracket, and remove trailing spaces
+    rawNames <- names(collectionDT)
+    names(collectionDT)[! names(collectionDT) %in% collectionIdColumns] <- sub("\\s*\\[([^\\[]*)$", "", names(collectionDT)[! names(collectionDT) %in% collectionIdColumns])
+    names(collectionDT)[names(collectionDT) == 'Incertae Sedis'] <- rawNames[names(collectionDT) == 'Incertae Sedis']
+
+    if (!!length(object@metadata@data)) {
+
+        # need to be sure sample metadata contains only the relevant rows, actually having assay data
+        # also need to make sure it has the assay record id column
+        sampleMetadataDT <- data.table::setDT(merge(
+            object@metadata@data, 
+            collectionDT[, collectionIdColumns, with = FALSE], 
+            by = c(object@metadata@ancestorIdColumns, object@metadata@recordIdColumn),
+            all.y = TRUE
+        ))
+
+        # if we only want continuous metadata, only keep numeric columns
+        # means we lose dates, but i think thats ok for now
+        if (continuousMetadataOnly) {
+            metadataColNames <- names(sampleMetadataDT)
+            numericColumns <- metadataColNames[which(sapply(sampleMetadataDT,is.numeric))]
+            sampleMetadataDT <- sampleMetadataDT[, unique(c(collectionIdColumns, numericColumns)), with=FALSE]
+        }
+
+        # also need to make sure they are in the same order
+        data.table::setorderv(sampleMetadataDT, cols=collection@recordIdColumn)
+        data.table::setorderv(collectionDT, cols=collection@recordIdColumn)
```

Comment:
are we sure all the same records still exist between these two? Could a record have gotten dropped from one or the other? Worth a check or a test?